<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
<head>
  <meta http-equiv="Content-Language" content="en-us">
  <meta http-equiv="Content-Type">
  <link rel="stylesheet" type="text/css" href="css/repminer.css">
  <title>RepMiner</title>
  <link rel="SHORTCUT ICON"
 href="file:///Users/jestill/code/repminer/trunk/html/jestill/JE5.ico">
</head>
<body>
<table style="border-collapse: collapse;" id="AutoNumber1" border="0"
 bordercolor="#111111" cellpadding="0" cellspacing="0" width="100%">
  <tbody>
    <tr>
      <td width="50%"> <img src="ap_cluster_files/RepMiner.png"
 border="1" height="115" width="326"></td>
      <td width="50%"> <big style="font-weight: bold;"><big><a
 class="mozTocH1" name="mozTocId974589"></a>RepMiner Process Manual<br>
      <small>Cytoscape Visualization</small></big></big><br>
      <br>
      <a href="http://jestill.myweb.uga.edu">James C. Estill</a><br>
November 7, 2008<br>
      </td>
    </tr>
  </tbody>
</table>
<hr>
<big><big><a style="font-weight: bold;" class="mozTocH2"
 name="mozTocId121997"></a><span style="font-weight: bold;">!! This
Document is Under Current Development !!<br>
<br>
</span></big></big><a class="mozTocH4" name="mozTocId722650"></a><span
 style="font-weight: bold;">This program manual and the scripts that
make up the RepMiner
package are under current development. Everything is subject to change
without notice at this point. This software comes as is, without any
expressed or implied warranty. Use at your own risk.</span><br
 style="font-weight: bold;">
<a style="font-weight: bold;" class="mozTocH4" name="mozTocId767722"></a><span
 style="font-weight: bold;">The RepMiner programs are released under
the </span><a style="font-weight: bold;"
 href="http://www.gnu.org/licenses/gpl.html">GNU License version 3</a><span
 style="font-weight: bold;">.
All other
software linked from this site are covered by the individual licenses
of the software authors. <br>
<br>
The basic develpment plan for RepMiner</span><br
 style="font-weight: bold;">
<br style="font-weight: bold;">
<ol style="font-weight: bold;">
  <li>First generation completely PERL based</li>
  <ol>
    <li>Does not make much use of database <br>
    </li>
    <li>Scripts directly move data from one format to another<br>
    </li>
  </ol>
  <li>BioSQL back end</li>
  <ol>
    <li>Add further phylogeny support to BioSQL</li>
    <li>Programs will have the option to dump the output to the BioSQL
database</li>
    <ol>
      <li>This must be left an option so that the text format option is
still available, and a database backend is not required.<br>
      </li>
    </ol>
  </ol>
  <li>HTML GUI</li>
  <ol>
    <li>Order that will be generaate</li>
    <ul>
      <li><br>
      </li>
    </ul>
  </ol>
  <ol>
    <li>Can start to do some of these processes with an HTML based GUI
interface</li>
    <li>These will be PERL cgi scripts that make heavy use of the
existing PERL code<br>
    </li>
    <li><br>
    </li>
  </ol>
  <li>Java interfaces</li>
  <ol>
    <li>This will benefit from the BioSQL back end.<br>
    </li>
  </ol>
</ol>
<span style="font-weight: bold;"><br>
</span>The following information relates to visualizing resulst using
the Cytoscape program.<br>
<br>
<hr style="width: 100%; height: 2px;">
<h1><br>
</h1>
Append a number to the beginning of each fasta record. This creates a
unique id that RepMiner will use in the database, and allows for an
easy way to draw the all by all blast matrix.<br>
<pre class="prettyprint"><span class="pun">  &gt;fasta_add_num.pl -i my_seqs.fasta -o my_seqs_num.fasta</span></pre>
<br>
Create a MySQL database to hold your records. Eventually this will be
switched to a more robust database framework based on BioSQL, but the
first version of RepMiner will stick with this.<br>
<br>
Log on to mysql:<br>
<pre class="prettyprint"><span class="pun">  &gt;mysql -u username -p</span></pre>
<br>
Create the database:<br>
<pre class="prettyprint"><span class="pun">  &gt;create database myseqs;<br></span><span
 class="pun">  &gt;exit;</span><br></pre>
The table structures will be created automatically when loading data to
the database. To load the sequence files to the database ..<br>
<br>
<pre class="prettyprint"><span class="pun">  &gt;mysql -u username -p</span></pre>
Load the fasta file to the database.<br>
<pre class="prettyprint"><span class="pun">  &gt;fasta2db.pl -i </span><span
 class="pun">my_seqs_num.fasta</span>
</pre>
Do the thing<br>
<pre class="prettyprint"><span class="pun">  &gt;jabablast.pl -i </span></pre>
<br>
<br>
The full process for uploadin<br>
<br>
<pre class="prettyprint"><span class="pun">  &gt;formatdb -i my_seqs.fasta -p F -t my_seqs -n my_seqs</span></pre>
<br>
<br>
Generate datatabse for all by all blast<br>
<br>
<pre class="prettyprint"><span class="pun">  &gt;formatdb -i my_seqs_num.fasta -p F -t my_seqs -n my_seqs</span></pre>
<br>
Do the all by all blast<br>
<pre class="prettyprint"><span class="pun">  &gt;formatdb -i my_seqs.fasta -p F -t my_seqs -n my_seqs</span></pre>
<br>
<br>
<br>
Blast the fasta file against a TE database of your choice ...<br>
<br>
<pre class="prettyprint"><span class="pun">  &gt;blastall -p blastn -i my_seqs.fasta -d TREP10 </span></pre>
<br>
<br>
<hr style="width: 100%; height: 2px;"><br>
Given an NCBI BLAST result in the tab delimited alignment format (-m 8)
you can create a sif format file for cytoscape using the following
command.<br>
<pre class="prettyprint"><span class="pun">  &gt;cnv_blast2sif.pl -i all_rpn_all_rpn.bln -o network -u jestill -f all_exemp -m 8<br></span></pre>
<br>
<h1><a class="mozTocH3" name="mozTocId629196"></a>References</h1>
<ul>
</ul>
<ol>
  <li>Brohee, S. and J. van Helden (2006). "Evaluation of clustering
algorithms for protein-protein interaction networks." BMC
Bioinformatics 7: 488.<br>
  </li>
  <li>Dongen, S. M. v. (2000). Graph Clustering by Flow Simulation.
Centre for Mathematics and Computer Science (CWI). Amsterdam,
University of Utrecht. PhD: 169.</li>
  <li>Enright, A. J., S. Van Dongen, et al. (2002). "An efficient
algorithm for large-scale detection of protein families." Nucleic Acids
Res 30(7): 1575-84.<br>
  </li>
  <li>Frey, B. J. and D. Dueck (2007). "Clustering by passing messages
between data points." Science 315(5814): 972-6. [<a
 href="http://www.psi.toronto.edu/affinitypropagation/FreyDueckScience07.pdf">PDF</a>]</li>
  <li>Frey, B. J. and D. Dueck (2008). "Response to comment on
"clustering by passing messages between data points"." Science
319(5864): -.</li>
  <li>Koenker, R. and P. Ng (2008). SparseM: A Sparse Matrix Package
for R.<br>
  </li>
</ol>
<ul>
</ul>
<hr>
<table style="border-collapse: collapse;" id="AutoNumber2" border="0"
 bordercolor="#111111" cellpadding="0" cellspacing="0" width="100%">
  <tbody>
    <tr>
      <td width="50%"><i>Contact: <a
 href="http://jestill.myweb.uga.edu">James
Estill</a><br>
Last Updated: November, 3 2008<!--webbot bot="Timestamp" i-checksum="60193" endspan --></i></td>
      <td width="50%"> <br>
      </td>
      <td width="50%">
      <p> <a href="http://sourceforge.net/"> <img
 src="ap_cluster_files/sflogo.png" alt="SourceForge.net Logo" border="0"
 height="62" width="210"> </a> </p>
      </td>
    </tr>
  </tbody>
</table>
<script type="text/javascript">
var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
</script>
<script src="ap_cluster_files/ga.js" type="text/javascript"></script>
<script type="text/javascript">
var pageTracker = _gat._getTracker("UA-3163301-1");
pageTracker._initData();
pageTracker._trackPageview();
</script>
</body>
</html>
